cucumber-opentelemetry/src/main/java/io/cucumber/opentelemetry/OpenTelemetryTracingEventListener.java

```diff
@@ -0,0 +1,289 @@
+                    .setAttribute(ATTRIBUTE_KEY_STEP_TYPE, step.getClass().getSimpleName());
+        }
+        startSpan(event, event.getTestStep().getCodeLocation(), spanBuilder);
+    }
+
+    private void onTestStepFinished(final TestStepFinished event) {
+        endSpan(event, event.getTestStep().getCodeLocation(), event.getResult());
+    }
+
+    private static String fileColonLine(TestCaseEvent event) {
+        return event.getTestCase().getUri().getSchemeSpecificPart() + ":" + event.getTestCase().getLocation().getLine();
+    }
+
+    private static String getHookDescription(TestStepStarted event) {
+        final Matcher m = CODE_LOCATION_CLASS_METHOD_PATTERN.matcher(event.getTestStep().getCodeLocation());
+        if (m.matches()) {
+            // Just the class name (without package) and method
+            return m.group(2) + "." + m.group(3);
+        } else {
+            return event.getTestStep().getCodeLocation();
+        }
+    }
+
+    private void startSpan(TestCaseEvent event, String codeLocation, SpanBuilder spanBuilder) {
+        spanBuilder.setAttribute(ATTRIBUTE_KEY_CUCUMBER_EVENT, event.getClass().getSimpleName());
+
+        final Matcher m = CODE_LOCATION_CLASS_METHOD_PATTERN.matcher(codeLocation);
+        if (m.matches()) {
+            spanBuilder.setAttribute(SemanticAttributes.CODE_NAMESPACE, m.group(1) + "." + m.group(2));
+            spanBuilder.setAttribute(SemanticAttributes.CODE_FUNCTION, m.group(3));
+            if (!m.group(4).isEmpty()) {
+                spanBuilder.setAttribute(ATTRIBUTE_KEY_CODE_FUNCTION_ARGS,
+                    Arrays.asList(m.group(4).split(",\\s*")));
+            }
+        } else {
+            spanBuilder.setAttribute(ATTRIBUTE_KEY_CODE_LOCATION, codeLocation);
+        }
+
+        final Span span = spanBuilder.startSpan();
+        final Scope scope = span.makeCurrent();
+
+        testCaseToSpanScope.put(Pair.of(event.getTestCase().getId(), codeLocation), Pair.of(span, scope));
+    }
+
+    private void endSpan(TestCaseEvent event, String codeLocation, Result result) {
+        final Span span = Span.current();
+
+        span.setAttribute(ATTRIBUTE_KEY_CUCUMBER_STATUS, result.getStatus().toString());
+
+        if (result.getStatus() == Status.FAILED) {
+            final Throwable error = result.getError();
+            span.recordException(error);
+            span.setStatus(StatusCode.ERROR, error.getMessage());
+        } else if (result.getStatus() == Status.PASSED) {
+            span.setStatus(StatusCode.OK);
+        }
+
+        final Pair<Span, Scope> spanScope = testCaseToSpanScope
+                .remove(Pair.of(event.getTestCase().getId(), codeLocation));
+        if (spanScope != null) {
+            if (DEBUG_SPAN_SCOPE) {
+                if (!span.equals(spanScope.getKey())) {
+                    System.err.println(getClass().getSimpleName() + ": " +
```

Comment:
Isn't ”System.err" a bit old fashioned? Maybe standard logging would be more appropriate.

Reply:
Oops, yes, very much! And I'm clearly old fashioned in this regard--at least in code I'm not familiar with. ;-) I'll take care of that.